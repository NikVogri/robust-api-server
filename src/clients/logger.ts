import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
