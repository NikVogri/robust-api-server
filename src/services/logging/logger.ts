import winston from 'winston';
import { AppError } from '../../error/AppError';
import { singleton } from 'tsyringe';
import { Request, Response } from '../../types';

@singleton()
export class Logger {
    public readonly winston: winston.Logger;

    private readonly custom = {
        levels: {
            error: 0,
            info: 1,
            http: 2,
            verbose: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
            verbose: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.winston = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'verbose',
                format: winston.format.combine(
                    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                    winston.format.printf(
                        (i) => `${i.level} ${i.timestamp}: ${i.message} ${JSON.stringify(i.stack) ?? ''}`,
                    ),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error) {
        this.winston.error(message, AppError.toStringifiableObject(error));
    }

    info(message: string) {
        this.winston.info(message);
    }

    http(req: Request, res: Response, error?: Error) {
        const log = this.buildHttpLog(req, res, error);
        this.winston.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: unknown): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: Object.keys(req.query).length > 0 ? req.query : undefined,
                params: Object.keys(req.params).length > 0 ? req.params : undefined,
                body: Object.keys(req.body).length > 0 ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: this.buildErrorObject(error),
        });
    }

    private buildErrorObject(error?: unknown) {
        if (!error || !(error instanceof Error)) return undefined;

        const output: Record<string, unknown> = {
            message: error?.message,
        };

        if (error instanceof AppError && error.thrownError) {
            output['thrown'] = AppError.toStringifiableObject(error.thrownError);
        }

        return output;
    }
}
