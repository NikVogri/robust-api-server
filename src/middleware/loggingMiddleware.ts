import { NextFunction, Request, Response } from 'express';
import { Logger } from '../clients/logger';
import { container } from 'tsyringe';

export const loggingMiddleware = (req: Request, res: Response, next: NextFunction) => {
    const logger = container.resolve(Logger);
    const end = res.end;

    // @ts-expect-error - overriding the res.end function which gets called last
    res.end = (chunk, encoding: BufferEncoding) => {
        res.end = end;
        res.end(chunk, encoding);

        // @ts-expect-error - TODO: fix Request type to include optional thrownError
        // In case request has an additional 'thrownError' property, include it in the log.
        logger.http(req, res, req.thrownError);
    };

    next();
};
