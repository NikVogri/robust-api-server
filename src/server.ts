import express, { Response } from 'express';
import { catsRouter } from './routes/cats';
import { errorMiddleware } from './middleware/errorMiddleware';
import { loggingMiddleware } from './middleware/loggingMiddleware';
import { Logger } from './clients/logger';
import { container } from 'tsyringe';

export const startApiServer = () => {
    const logger = container.resolve(Logger);
    const app = express();

    // Set middleware
    app.use(express.json());
    app.use(loggingMiddleware);

    // Set HTTP headers

    // Set routers
    app.use(catsRouter);

    // Set debug endpoints
    app.get('/status', (_, res: Response) => {
        res.status(200).send({ time: Date.now() });
    });

    // Set error middleware
    app.use(errorMiddleware);

    const PORT = process.env.PORT;
    return app.listen(PORT, () => logger.info('Server started listening on port ' + PORT));
};
