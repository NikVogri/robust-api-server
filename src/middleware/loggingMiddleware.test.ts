import { afterAll, describe, expect, it, vi } from 'vitest';
import { loggingMiddleware } from './loggingMiddleware';
import { container } from 'tsyringe';

import { getRequest, getResponse } from '../../vitest.mock';
import { Logger } from '../services/logging/logger';
import { startApiServer } from '../server';
import supertest from 'supertest';

vi.mock('../services/logging/logger', () => {
    const Logger = vi.fn(() => ({
        http: vi.fn(),
        info: vi.fn(),
    }));

    return { Logger };
});

describe('Test loggingMiddleware', () => {
    container.registerInstance(Logger, new Logger());
    const logger = container.resolve(Logger);

    afterAll(() => {
        container.reset();
    });

    it('should call the next function', async () => {
        expect.assertions(1);

        const request = getRequest();
        const response = getResponse();
        const nextFn = vi.fn();

        loggingMiddleware(request, response, nextFn);
        expect(nextFn).toHaveBeenCalled();
    });

    it('should log the request before sending a response', async () => {
        expect.assertions(2);

        vi.spyOn(logger, 'http').mockImplementationOnce((...args) => {
            expect(args[0].path).toMatchSnapshot();
            expect(args[1].statusCode).toMatchSnapshot();

            return undefined;
        });

        const app = startApiServer();
        await supertest(app).get(`/status`);
        app.close();
    });
});
