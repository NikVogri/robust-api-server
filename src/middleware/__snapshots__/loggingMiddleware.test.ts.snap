// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test loggingMiddleware > should log the request before sending a response 1`] = `"/status"`;

exports[`Test loggingMiddleware > should log the request before sending a response 2`] = `200`;
